src/webob/acceptparse.py

```diff
@@ -268,6 +678,1197 @@ def _match(self, mask, item):
+
+        if isinstance(
+            other, (AcceptLanguageNoHeader, AcceptLanguageInvalidHeader)
+        ):
+            return AcceptLanguageNoHeader()
+
+        return self._add_instance_and_non_accept_language_type(
+            instance=self, other=other,
+        )
+
+    def __radd__(self, other):
+        """
+        Add to header, creating a new header object.
+
+        See the docstring for :meth:`AcceptLanguageValidHeader.__add__`.
+        """
+        return self._add_instance_and_non_accept_language_type(
+            instance=self, other=other, instance_on_the_right=True,
+        )
+
+    def __repr__(self):
+        return '<{}>'.format(self.__class__.__name__)
+        # We do not display the header_value, as it is untrusted input. The
+        # header_value could always be easily obtained from the .header_value
+        # property.
+
+    def __str__(self):
+        """Return the ``str`` ``'<invalid header value>'``."""
+        return '<invalid header value>'
+
+    def _add_instance_and_non_accept_language_type(
+        self, instance, other, instance_on_the_right=False,
+    ):
+        if not other:
+            return AcceptLanguageNoHeader()
+
+        other_header_value = self.python_value_to_header_str(value=other)
+
+        try:
+            return AcceptLanguageValidHeader(header_value=other_header_value)
+        except ValueError:  # invalid header value
+            return AcceptLanguageNoHeader()
+
+
+def create_accept_language_header(header_value):
+    """
+    Create an object representing the ``Accept-Language`` header in a request.
+
+    :param header_value: (``str``) header value
+    :return: If `header_value` is ``None``, an :class:`AcceptLanguageNoHeader`
+             instance.
+
+             | If `header_value` is a valid ``Accept-Language`` header, an
+               :class:`AcceptLanguageValidHeader` instance.
+
+             | If `header_value` is an invalid ``Accept-Language`` header, an
+               :class:`AcceptLanguageInvalidHeader` instance.
+    """
+    if header_value is None:
+        return AcceptLanguageNoHeader()
+    try:
+        return AcceptLanguageValidHeader(header_value=header_value)
+    except ValueError:
+        return AcceptLanguageInvalidHeader(header_value=header_value)
+
+
+def accept_language_property():
+    doc = """
+        Property representing the ``Accept-Language`` header.
+
+        (:rfc:`RFC 7231, section 5.3.5 <7231#section-5.3.5>`)
+
+        The header value in the request environ is parsed and a new object
+        representing the header is created every time we *get* the value of the
+        property. (*set* and *del* change the header value in the request
+        environ, and do not involve parsing.)'
+    """
+
+    ENVIRON_KEY = 'HTTP_ACCEPT_LANGUAGE'
+
+    def fget(request):
+        """Get an object representing the header in the request."""
+        return create_accept_language_header(
```

Comment:
Here we return a brand new `AcceptLanguage*` object each time, yet when we assign a new `AcceptLanguage` using fset we set it directly to the AcceptLanguage object.

Would it make sense here to check if it is a string, and if so then run the `create_accept_language_header` function, followed by setting the environ key to the newly created object? This way if a user calls `request.accept_language` twice they only have the overhead of creating the object the first time? If they modify `request.environ` directly it would no longer be an `AcceptLanguage` object and would correctly re-create it.

----

Another thing I noticed is that with your new `__add__` functions since they don't modify the existing one but create a new one something like this is not possible, or am I mistaken:

```
request.accept_langauge += 'nl-NL;q=0.7'
```

Since that would return a new object and `request.accept_language` would be untouched. I can live with that... just not what I expect at the moment based upon the other properties.